File: go/2392. Build a Matrix With Conditions.go
func buildMatrix(k int, rowConditions [][]int, colConditions [][]int) [][]int {
    rowOrder := getOrder(getGraph(rowConditions, k), k)
    colOrder := getOrder(getGraph(colConditions, k), k)
    if len(rowOrder) == 0 || len(colOrder) == 0 {
        return [][]int{}
    }
    answer := make([][]int, k)
    for i := range answer {
        answer[i] = make([]int, k)
    }
    for i := 1; i <= k; i++ {
        answer[rowOrder[i]][colOrder[i]] = i
    }
    return answer
}

func getGraph(edges [][]int, k int) map[int][]int {
    graph := make(map[int][]int)
    for _, edge := range edges {
        from, to := edge[0], edge[1]
        if _, ok := graph[from]; !ok {
            graph[from] = []int{}
        }
        graph[from] = append(graph[from], to)
    }
    return graph
}

func getOrder(graph map[int][]int, k int) []int {
    inDegrees := make([]int, k+1)
    for i := 1; i <= k; i++ {
        if _, ok := graph[i]; !ok { continue }
        for _, x := range graph[i] {
            inDegrees[x]++
        }
    }
    q := []int{}
    for i := 1; i <= k; i++ {
        if inDegrees[i] == 0 {
            q = append(q, i)
        }
    }
    order := []int{}
    for len(q) > 0 {
        current := q[0]
        q = q[1:]
        order = append(order, current)
        if _, ok := graph[current]; !ok { continue }
        for _, next := range graph[current] {
            inDegrees[next]--
            if inDegrees[next] == 0 {
                q = append(q, next)
            }
        }
    }
    if len(order) != k {
        return []int{}
    }
    result := make([]int, k+1)
    index := 0
    for _, i := range order {
        result[i] = index
        index++
    }
    return result
}

File: java/2392. Build a Matrix With Conditions.java
class Solution {
    public int[][] buildMatrix(int k, int[][] rowConditions, int[][] colConditions) {
        int[] rowOrder = getOrder(getGraph(rowConditions, k), k);
        int[] colOrder = getOrder(getGraph(colConditions, k), k);
        if (rowOrder.length == 0 || colOrder.length == 0) {
            return new int[0][0];
        }
        int[][] answer = new int[k][k];
        for (int i = 1; i <= k; i++) {
            answer[rowOrder[i]][colOrder[i]] = i;
        }
        return answer;
    }
    private Map<Integer, List<Integer>> getGraph(int[][] edges, int k) {
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int[] edge : edges) {
            int from = edge[0], to = edge[1];
            graph.computeIfAbsent(from, key -> new ArrayList<>()).add(to);
        }
        return graph;
    }

    private int[] getOrder(Map<Integer, List<Integer>> graph, int k) {
        int[] inDegrees = new int[k+1];
        for (int i = 1; i <= k; i++) {
            if (!graph.containsKey(i)) continue;
            for (int x : graph.get(i)) inDegrees[x]++;
        }
        Deque<Integer> q = new ArrayDeque<>();
        for (int i = 1; i <= k; i++) {
            if (inDegrees[i] == 0) q.addLast(i);
        }
        List<Integer> order = new ArrayList<>();
        while (!q.isEmpty()) {
            int current = q.pollFirst();
            order.add(current);
            if (!graph.containsKey(current)) continue;
            for (int next : graph.get(current)) {
                inDegrees[next]--;
                if (inDegrees[next] == 0) q.addLast(next);
            }
        }
        if (order.size() != k) return new int[0];
        int[] result = new int[k+1];
        int index = 0;
        for (int i : order) {
            result[i] = index++;
        }
        return result;
    }
}